### PR TITLE
Add KeyMatch4 function support

### LINKED ISSUE #61: DynamicExpresso.Exceptions.UnknownIdentifierException: 'Unknown identifier 'keyMatch4' (at index 65).
KeyMatch4 is not supported, can we pls add the support?

### PR DESCRIPTION
**Pull request by Sagilio**
close #61

### CODE REVIEW

**Comment by hsluoyz** [comment]
@mitesh-diligencevault can you review this PR?

File: NetCasbin.UnitTest/BuiltInOperatorTest.cs [hunk 1]
```diff
@@ -12,5 +13,27 @@ public void TestIpMatch()
...
+            new object[] {"/parent/123/child/123", "/parent/{id}/child/{id}", true},
+            new object[] {"/parent/123/child/456", "/parent/{id}/child/{id}", false},
+            new object[] {"/parent/123/child/123", "/parent/{id}/child/{another_id}", true},
+            new object[] {"/parent/123/child/456", "/parent/{id}/child/{another_id}", true},
+            new object[] {"/parent/123/child/123/book/123", "/parent/{id}/child/{id}/book/{id}", true},
+            new object[] {"/parent/123/child/123/book/456", "/parent/{id}/child/{id}/book/{id}", false},
+            new object[] {"/parent/123/child/456/book/123", "/parent/{id}/child/{id}/book/{id}", false},
+            new object[] {"/parent/123/child/456/book/", "/parent/{id}/child/{id}/book/{id}", false},
+            new object[] {"/parent/123/child/456", "/parent/{id}/child/{id}/book/{id}", false},
+            new object[] {"/parent/123/child/123", "/parent/{i/d}/child/{i/d}", false}
```
Comment (mitesh-diligencevault) [inline NetCasbin.UnitTest/BuiltInOperatorTest.cs:28]: what does this mean `{i/d}`?

Comment (Sagilio) [reply NetCasbin.UnitTest/BuiltInOperatorTest.cs:28]: This case wants to test routes similar to expressions param should not be misjudged. The correct case is the first case of all.

File: NetCasbin/Util/BuiltInFunctions.cs [hunk 2]
```diff
@@ -81,6 +86,64 @@ public static bool KeyMatch3(string key1, string key2)
...
+
+            var valueRegex = new Regex($"^{key2}$");
+            var key1Match =  valueRegex.Match(key1);
+
+            if (!key1Match.Success)
+            {
+                return false;
+            }
+
+            int key1GroupMatchCount = key1Match.Groups.Count - 1;
```
Comment (huazhikui) [inline NetCasbin/Util/BuiltInFunctions.cs:121]: my implement:
```c#
public static bool KeyMatch4(string key1, string key2)
{
    key2 = key2.Replace("/*", "/.*");

    var tokens = new Dictionary<string, int>();

    var regex = new Regex("\\{([^/]+)\\}");
    key2 = regex.Replace(key2, (matched) =>
    {
        string token = matched.Value.Substring(1, matched.Value.Length - 2);
        if (!tokens.ContainsKey(token))
        {
            tokens.Add(token, 0);
        }
        tokens[token]++;
        return $"(?'{token}'[^/]+)";
    });

    regex = new Regex($"^{key2}$");
    Match key1Matched = regex.Match(key1);

    if (!key1Matched.Success)
    {
        return false;
    }

    int key1GroupMatchCount = key1Matched.Groups.Count - 1;
    if (key1GroupMatchCount != tokens.Count)
    {
        throw new ArgumentException("KeyMatch4: number of tokens is not equal to number of values");
    }

    foreach (var tokenPair in tokens)
    {
        Group tokenGroup = key1Matched.Groups[tokenPair.Key];
        if (!tokenGroup.Success)
        {
            return false;
        }
        IEnumerable<string> captures = tokenGroup.Captures.Cast<Capture>().Select(x => x.Value);
        string first = captures.First();
        if (captures.Any(x => x != first))
        {
            return false;
        }
    }
    return true;
}
```

Comment (Sagilio) [reply NetCasbin/Util/BuiltInFunctions.cs:121]: I have tried to run this implement as the KeyMatch4New, It looks will more memory allocated and not well performance than now.
|       Method |        Job |       Runtime |                 key1 |                 key2 |     Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |----------- |-------------- |--------------------- |--------------------- |---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|    **KeyMatch4** | **Job-SZGZJN** |      **.NET 4.8** | **/pare(...)d/123 [21]** | **/pare(...)r_id} [31]** | **19.03 μs** | **1.018 μs** | **0.673 μs** |  **1.19** |    **0.04** | **3.4180** |     **-** |     **-** |  **10.52 KB** |
|    KeyMatch4 | Job-ZPYDIF | .NET Core 3.1 | /pare(...)d/123 [21] | /pare(...)r_id} [31] | 16.05 μs | 0.549 μs | 0.363 μs |  1.00 |    0.00 | 2.7161 |     - |     - |   8.37 KB |
|              |            |               |                      |                      |          |          |          |       |         |        |       |       |           |
| KeyMatch4New | Job-SZGZJN |      .NET 4.8 | /pare(...)d/123 [21] | /pare(...)r_id} [31] | 26.64 μs | 3.925 μs | 2.596 μs |  1.11 |    0.11 | 3.9063 |     - |     - |  12.06 KB |
| KeyMatch4New | Job-ZPYDIF | .NET Core 3.1 | /pare(...)d/123 [21] | /pare(...)r_id} [31] | 24.00 μs | 1.432 μs | 0.947 μs |  1.00 |    0.00 | 3.1128 |     - |     - |   9.72 KB |
|              |            |               |                      |                      |          |          |          |       |         |        |       |       |           |
|    **KeyMatch4** | **Job-SZGZJN** |      **.NET 4.8** | **/pare(...)d/123 [21]** | **/pare(...)/{id} [23]** | **20.10 μs** | **0.561 μs** | **0.371 μs** |  **1.15** |    **0.05** | **3.3875** |     **-** |     **-** |  **10.45 KB** |
|    KeyMatch4 | Job-ZPYDIF | .NET Core 3.1 | /pare(...)d/123 [21] | /pare(...)/{id} [23] | 17.50 μs | 0.792 μs | 0.524 μs |  1.00 |    0.00 | 2.6855 |     - |     - |    8.3 KB |
|              |            |               |                      |                      |          |          |          |       |         |        |       |       |           |
| KeyMatch4New | Job-SZGZJN |      .NET 4.8 | /pare(...)d/123 [21] | /pare(...)/{id} [23] | 23.94 μs | 0.686 μs | 0.454 μs |  1.18 |    0.05 | 3.5706 |     - |     - |     11 KB |
| KeyMatch4New | Job-ZPYDIF | .NET Core 3.1 | /pare(...)d/123 [21] | /pare(...)/{id} [23] | 20.32 μs | 1.032 μs | 0.683 μs |  1.00 |    0.00 | 2.8076 |     - |     - |   8.69 KB |

**Pull request merged**